### ISSUE TITLE
Build an `ammonite-compilation` optional submodule to support compiler/type/bytecode twiddling

### ISSUE THREAD

**Issue by lihaoyi** [lihaoyi/Ammonite#286]
Various features related to fiddling with the compiler, fancy types, or the runtime have been requested over time
- [Support for :kind](https://gitter.im/lihaoyi/Ammonite?at=56618a5b77d271e66db1eb95)
- [Support for :power-mode](https://github.com/lihaoyi/Ammonite/issues/167)
- [Support for :javap](https://github.com/lihaoyi/Ammonite/issues/272)

None of these are IMO general enough to live in the core REPL. If I'm using Ammonite as an EC2 management shell, or as a small-data graph-exploration shell, or as a filesystem shell, these things are totally out of scope. 

Nevertheless, these would be handy for compiler geeks, which the Scala community is full of! Thus, although people an already piecemeal add things ad-hoc to their own `predef.scala` files, it would probably be useful to bundle a few of them together into a `ammonite-compilation` submodule, similar to `ammonite-shell`. 

This would probably be popular enough to live inside the Ammonite repo with the same status as `ammonite-shell`: published together, and documented together, but pulled in separately only for people who want it via an incantation they would place in their `predef.scala` file.

**Comment by lihaoyi** [comment]
Here's a sample implementation of `:kind`

Usage:
```scala
ammonite-repl> kind[Option[Int]]
res1: String = "Option's kind is * -> *. This is a type constructor: a 1st-order-kinded type."
ammonite-repl> kind[Map[_,_]]
res2: String = "Map's kind is * -> * -> *. This is a type constructor: a 1st-order-kinded type."
```

Definition

```scala
object Catnip {
  implicit class IdOp[A](val a: A) extends AnyVal {
    def some: Option[A] = Some(a)
  }
  def none[A]: Option[A] = None
  def kind[A: scala.reflect.runtime.universe.TypeTag]: String = {
    import scala.reflect.runtime.universe.{typeOf => _, _}
    def typeKind(sig: Type): String = sig match {
      case PolyType(params, resultType) =>
        (params map { p =>
          typeKind(p.typeSignature) match {
            case "*" => "*"
            case s   => "(" + s + ")"
          }
        }).mkString(" -> ") + " -> *"
      case _ => "*"
    }
    def typeSig(tpe: Type): Type = tpe match {
      case SingleType(pre, sym) => sym.companionSymbol.typeSignature
      case ExistentialType(q, TypeRef(pre, sym, args)) => sym.typeSignature
      case TypeRef(pre, sym, args) => sym.typeSignature
    }
    val sig = typeSig(typeOf[A])
    val s = typeKind(sig)
    sig.typeSymbol.name + "'s kind is " + s + ". " + (s match {
      case "*" =>
        "This is a proper type."
      case x if !(x contains "(") =>
        "This is a type constructor: a 1st-order-kinded type."
      case x =>
        "This is a type constructor that takes type constructor(s): a higher-kinded type."
    })
  }
}
import Catnip._
```

Taken from

https://gitter.im/lihaoyi/Ammonite?at=5662b56977d271e66db210d8

I have no idea how correct this is but someone else probably does

**Comment by paulp** [comment]
The bar for `:kind` is [very very low](https://issues.scala-lang.org/browse/SI-8529) and I'm surprised there are people asking for it. I am doubtful they understand the answers they get.

Look how diplomatic I was. "Difficult to justify." Ha.

**Comment by lihaoyi** [comment]
If nobody implemented this after 1.5 years I'm going to call this a wontfix
